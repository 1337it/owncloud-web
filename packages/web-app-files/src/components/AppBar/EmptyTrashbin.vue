<template>
  <action-menu-item :action="emptyTrashBinAction" :items="selectedFiles" appearance="outline" />
</template>

<script>
import { mapGetters } from 'vuex'
import ActionMenuItem from '../ActionMenuItem.vue'

import EmptyTrashBin from '../../mixins/actions/emptyTrashBin'

export default {
  name: 'EmptyTrashbin',
  components: { ActionMenuItem },
  mixins: [EmptyTrashBin],
  computed: {
    ...mapGetters('Files', ['selectedFiles']),

    emptyTrashBinAction() {
      return this.$_emptyTrashBin_items[0]
    }
  }
}
</script>
