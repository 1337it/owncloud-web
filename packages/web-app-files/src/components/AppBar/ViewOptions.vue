<template>
  <div>
    <oc-button
      id="files-view-options-btn"
      key="files-view-options-btn"
      data-testid="files-view-options-btn"
      :aria-label="viewButtonAriaLabel"
      variation="passive"
      appearance="raw"
      size="small"
      gap-size="xsmall"
    >
      <oc-icon name="tune" size="small" />
      <translate>View</translate>
    </oc-button>
    <oc-drop
      drop-id="files-view-options-drop"
      toggle="#files-view-options-btn"
      mode="click"
      class="uk-width-auto"
    >
      <oc-list>
        <li class="files-view-options-list-item">
          <oc-switch
            v-model="hiddenFilesShownModel"
            data-testid="files-switch-hidden-files"
            :label="$gettext('Show hidden files')"
          />
        </li>
        <li class="files-view-options-list-item">
          <oc-page-size
            v-model="$_filesListPagination_pageItemsLimit"
            data-testid="files-pagination-size"
            :label="$gettext('Items per page')"
            :options="[100, 500, 1000, $gettext('All')]"
            class="files-pagination-size"
          />
        </li>
      </oc-list>
    </oc-drop>
  </div>
</template>

<script>
import { mapMutations, mapState } from 'vuex'

import MixinFilesListPagination from '../../mixins/filesListPagination'

export default {
  mixins: [MixinFilesListPagination],

  computed: {
    ...mapState('Files', ['areHiddenFilesShown']),

    viewButtonAriaLabel() {
      return this.$gettext('Display customization options of the files list')
    },

    hiddenFilesShownModel: {
      get() {
        return this.areHiddenFilesShown
      },

      set(value) {
        this.SET_HIDDEN_FILES_VISIBILITY(value)
      }
    }
  },

  methods: {
    ...mapMutations('Files', ['SET_HIDDEN_FILES_VISIBILITY'])
  }
}
</script>

<style lang="scss" scoped>
.files-view-options-list-item {
  & > * {
    display: flex;
    justify-content: space-between;
  }

  & + & {
    margin-top: var(--oc-space-small);
  }
}
</style>
