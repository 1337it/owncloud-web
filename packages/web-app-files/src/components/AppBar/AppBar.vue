<template>
  <div id="files-app-bar">
    <oc-hidden-announcer :announcement="selectedResourcesAnnouncement" level="polite" />
    <div class="files-topbar oc-py-s">
      <h1 class="oc-invisible-sr" v-text="pageTitle" />
      <div class="oc-flex oc-flex-between">
        <oc-breadcrumb
          v-if="breadcrumbs.length"
          id="files-breadcrumb"
          data-testid="files-breadcrumbs"
          class="oc-flex oc-flex-middle"
          context-menu-padding="small"
          :items="breadcrumbs"
        >
          <template v-if="breadcrumbsContextActionsDisplayed" #contextMenu>
            <context-actions :items="breadcrumbsContextActionsItems" />
          </template>
        </oc-breadcrumb>
        <shares-navigation v-if="hasSharesNavigation" />
        <view-options />
      </div>
      <div class="files-app-bar-actions">
        <div class="oc-flex-1 oc-flex oc-flex-start" style="gap: 15px">
          <slot v-if="selectedFiles.length === 0" name="actions" />
          <size-info v-if="hasBulkActions && selectedFiles.length > 0" class="oc-visible@l" />
          <batch-actions v-if="hasBulkActions && selectedFiles.length > 0" />
        </div>
      </div>
      <slot name="static" />
    </div>
  </div>
</template>

<script>
import { mapGetters, mapState, mapMutations } from 'vuex'

import Mixins from '../../mixins'
import MixinFileActions from '../../mixins/fileActions'
import { isLocationSpacesActive } from '../../router'
import { useActiveLocation } from '../../composables'

import BatchActions from './SelectedResources/BatchActions.vue'
import ContextActions from '../FilesList/ContextActions.vue'
import SharesNavigation from './SharesNavigation.vue'
import SizeInfo from './SelectedResources/SizeInfo.vue'
import ViewOptions from './ViewOptions.vue'

export default {
  components: {
    BatchActions,
    ContextActions,
    SharesNavigation,
    SizeInfo,
    ViewOptions
  },
  mixins: [Mixins, MixinFileActions],
  props: {
    breadcrumbs: { type: Array, default: () => [] },
    breadcrumbsContextActionsDisplayed: { type: Boolean, default: false },
    breadcrumbsContextActionsItems: { type: Array, default: () => [] },
    hasBulkActions: { type: Boolean, default: false },
    hasSharesNavigation: { type: Boolean, default: false }
  },
  setup() {
    return {
      isPersonalLocation: useActiveLocation(isLocationSpacesActive, 'files-spaces-personal-home'),
      isSpacesProjectLocation: useActiveLocation(isLocationSpacesActive, 'files-spaces-project')
    }
  },
  data: () => ({
    newFileAction: null,
    path: '',
    fileFolderCreationLoading: false
  }),
  computed: {
    ...mapGetters(['getToken', 'capabilities', 'configuration', 'newFileHandlers', 'user']),
    ...mapGetters('Files', ['files', 'currentFolder', 'selectedFiles', 'publicLinkPassword']),
    ...mapState('Files', ['areHiddenFilesShown']),

    // we should generalize the usage of pageTitle throughout the project next
    pageTitle() {
      const title = this.$route.meta.title
      return this.$gettext(title)
    },
    areDefaultActionsVisible() {
      return this.selectedFiles.length < 1
    },

    selectedResourcesAnnouncement() {
      if (this.selectedFiles.length === 0) {
        return this.$gettext('No items selected.')
      }
      const translated = this.$ngettext(
        '%{ amount } item selected. Actions are available above the table.',
        '%{ amount } items selected. Actions are available above the table.',
        this.selectedFiles.length
      )
      return this.$gettextInterpolate(translated, { amount: this.selectedFiles.length })
    }
  },

  created() {
    // Storage returns a string so we need to convert it into a boolean
    const areHiddenFilesShown = window.localStorage.getItem('oc_hiddenFilesShown') || 'false'
    const areHiddenFilesShownBoolean = areHiddenFilesShown === 'true'

    if (areHiddenFilesShownBoolean !== this.areHiddenFilesShown) {
      this.SET_HIDDEN_FILES_VISIBILITY(areHiddenFilesShownBoolean)
    }
  },

  methods: {
    ...mapMutations('Files', ['SET_HIDDEN_FILES_VISIBILITY'])
  }
}
</script>

<style lang="scss" scoped>
#files-app-bar {
  background-color: var(--oc-color-background-default);
  box-sizing: border-box;
  z-index: 2;
  padding: 0 var(--oc-space-medium);
  border-top-right-radius: 15px;

  .files-app-bar-actions {
    align-items: center;
    display: flex;
    gap: var(--oc-space-small);
    justify-content: flex-end;
    min-height: 3rem;
  }

  #files-breadcrumb {
    min-height: 2rem;
  }
}
</style>
