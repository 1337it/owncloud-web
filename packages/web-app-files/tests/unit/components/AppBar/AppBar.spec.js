import { createLocalVue, shallowMount } from '@vue/test-utils'
import Files from '@/__fixtures__/files'
import Vuex from 'vuex'
import DesignSystem from 'owncloud-design-system'
import GetTextPlugin from 'vue-gettext'

import AppBar from '@files/src/components/AppBar/AppBar.vue'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(DesignSystem)
localVue.use(GetTextPlugin, {
  translations: 'does-not-matter.json',
  silent: true
})

const selectedFile = Files['/'][1]
const selectedFiles = [Files['/'][1], Files['/'][4]]

describe('AppBar component', () => {
  describe('with no selected files', () => {
    const store = createStore({ selected: [] })
    const wrapper = getShallowWrapper(store)

    it('renders basic appBar with content slot', () => {
      const slotEl = wrapper.find('ruby')
      expect(slotEl.exists()).toBeTruthy()

      expect(wrapper).toMatchSnapshot()
    })
  })
  describe('with one selected file', () => {
    it('tbd', () => {
      const store = createStore({ selected: selectedFile })
      const wrapper = getShallowWrapper(store)
      expect(wrapper).toMatchSnapshot()
    })
  })
  describe('with multiple selected files', () => {
    it('tbd', () => {
      const store = createStore({ selected: selectedFiles })
      const wrapper = getShallowWrapper(store)
      expect(wrapper).toMatchSnapshot()
    })
  })

  it.todo('renders breadcrumbs')
  it.todo('last breadcrumb item emits event upon click')
})

function getShallowWrapper(store = {}) {
  return shallowMount(AppBar, {
    localVue,
    mocks: {
      $route: {
        meta: {
          title: 'ExampleTitle'
        }
      }
    },
    propsData: {
      slots: {
        actions: '<button>Click me</button>',
        content: '<ruby>Static content</ruby>'
      }
    },
    store
  })
}

function createStore(state = { selected: [] }) {
  return new Vuex.Store({
    modules: {
      Files: {
        namespaced: true,
        state,
        getters: {
          selectedFiles: () => state.selected
        },
        mutations: {
          SET_HIDDEN_FILES_VISIBILITY: jest.fn()
        }
      }
    }
  })
}
