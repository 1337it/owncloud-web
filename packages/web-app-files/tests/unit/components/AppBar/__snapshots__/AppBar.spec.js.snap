// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBar component with multiple selected files tbd 1`] = `
<div id="files-app-bar" slots="[object Object]">
  <oc-hidden-announcer-stub level="polite" announcement="2 items selected. Actions are available above the table."></oc-hidden-announcer-stub>
  <div class="files-topbar oc-py-s">
    <h1 class="oc-invisible-sr">ExampleTitle</h1>
    <div class="oc-flex oc-flex-right">
      <!---->
      <!---->
      <div class="oc-flex">
        <view-options-stub></view-options-stub>
        <sidebar-toggle-stub></sidebar-toggle-stub>
      </div>
    </div>
    <div class="files-app-bar-actions">
      <div class="oc-flex-1 oc-flex oc-flex-start" style="gap: 15px;">
        <!---->
        <!---->
        <!---->
      </div>
    </div>
  </div>
</div>
`;

exports[`AppBar component with no selected files renders basic appBar with static slot 1`] = `
<div id="files-app-bar" slots="[object Object]">
  <oc-hidden-announcer-stub level="polite" announcement="No items selected."></oc-hidden-announcer-stub>
  <div class="files-topbar oc-py-s">
    <h1 class="oc-invisible-sr">ExampleTitle</h1>
    <div class="oc-flex oc-flex-between">
      <!---->
      <!---->
      <view-options-stub></view-options-stub>
    </div>
    <div class="files-app-bar-actions">
      <div class="oc-flex-1 oc-flex oc-flex-start" style="gap: 15px;">
        <!---->
        <!---->
      </div>
    </div>
  </div>
</div>
`;

exports[`AppBar component with one selected file tbd 1`] = `
<div id="files-app-bar" slots="[object Object]">
  <oc-hidden-announcer-stub level="polite" announcement="amount items selected. Actions are available above the table."></oc-hidden-announcer-stub>
  <div class="files-topbar oc-py-s">
    <h1 class="oc-invisible-sr">ExampleTitle</h1>
    <div class="oc-flex oc-flex-right">
      <!---->
      <!---->
      <div class="oc-flex">
        <view-options-stub></view-options-stub>
        <sidebar-toggle-stub></sidebar-toggle-stub>
      </div>
    </div>
    <div class="files-app-bar-actions">
      <div class="oc-flex-1 oc-flex oc-flex-start" style="gap: 15px;">
        <!---->
        <!---->
        <!---->
      </div>
    </div>
  </div>
</div>
`;
